### ISSUE TITLE
wrong timestamp in footer when browsing old revision

### ISSUE THREAD

**Issue by phkehl** [kensanata/oddmuse#1]
When browsing an older revision of a topic (action=browse id=... revision=...) the timestamp display in the footer changes from "Last edited ..." to "Edited ...", which makes me assume it should display that revision's timestamp. However, it seems to always (for any revision=...) display the latest revision's timestamp.

The information must be available, though. The history (action=history id=...) displays the timestamp of each revision correctly.

Unfortunately my Perl-foo is not strong enough to suggest a patch. I can see GetFooterTimestamp() using $Page{ts}, but I have no clue where the {ts} field is filled-in.

I might try again finding the problem, but until then I'll leave the issue here.

**Comment by kensanata** [comment]
I think the problem is that when we're looking at an old revision on BrowsePage, GetTextRevision returns the old text and the old summary, without changing %Page itself. PrintFooter then calls the functions on @MyFooters including GetFooterTimestamp which looks at %Page.

The question is whether GetTextRevision can simply set %Page or not…

**Comment by AlexDaniel** [comment]
> The question is whether GetTextRevision can simply set %Page or not…

I don't think so. Look: 

``` perl
sub GetTextRevision {
  …
  my %keep = GetKeptRevision($revision);
  …
  return ($keep{text}, $revision, $keep{summary});
}
```

In other words, `GetTextRevision` retrieves full **%Page** object, but does not return it. Instead, it returns only some bits of it.

I think that we should change `GetTextRevision` to return full object. There are some modules that we have to fix though:

``` bash
grep 'GetTextRevision' modules/*.pl
```

``` text
modules/ban-contributors.pl:    my ($old) = GetTextRevision($Page{revision}-1, 1);
modules/despam.pl:    my ($text, $rev) = GetTextRevision($revision, 1); # quiet
modules/namespaces.pl:  my ($text, $revision) = GetTextRevision(GetParam('revision', ''), 1);
modules/private-pages.pl:*OldPrivatePagesGetTextRevision = \&GetTextRevision;
modules/private-pages.pl:*GetTextRevision = \&NewPrivatePagesGetTextRevision;
modules/private-pages.pl:sub NewPrivatePagesGetTextRevision {
modules/private-pages.pl:  my ($text, $revision) = OldPrivatePagesGetTextRevision(@_);
modules/static-hybrid.pl:    my ($text, $rev) = GetTextRevision($revision, 1); # quiet
modules/svg-edit.pl:    ($data) = GetTextRevision($revision); # ignore revision reset
modules/thumbs.pl:     my ($text, $revision) = GetTextRevision(GetParam('revision', '')); # maybe revision reset!
```

**Comment by kensanata** [comment]
Or we could write two subs – and the new one that sets %Page is only called from BrowsePage…

**Comment by AlexDaniel** [comment]
You think that there are many site-specific modules that rely on `GetTextRevision`? If yes, then indeed, it is a good idea. However, I'd much rather refactor the whole code base (mainly because I think that there are many other places where an object should be returned, and I don't want to keep lots of compatibility subs).

**Comment by phkehl** [comment]
Thanks for the quick replies!

Couldn't you just make GetTextRevision() return the full object as a fourth item? It seems it's always called in list context and it should be transparent to the current uses.

Then pass the additional proper timestamp in BrowsePage() via PrintFooter() to GetFooterTimestamp().

(Not sure if I got the call tree right, and if this fits your concepts and Perl dialect.)

**Comment by AlexDaniel** [comment]
@phkehl sure! There are many ways to solve this problem. We can even pass a timestamp as fourth item, that could fix this bug as well.
Lately we were pushing for better code. Introducing one more item to the returned list is more like a move into a wrong direction. I'd much rather move forward even if we have to sacrifice just a tiny bit of compatibility.

**Comment by AlexDaniel** [comment]
Working on it!

**Comment by AlexDaniel** [comment]
Underlying issue was added to [Refactoring](https://oddmuse.org/wiki/Refactoring#returning_hashes) page. Comments are appreciated.

I have a bunch of failing tests, so I'm not rushing to commit it yet (even to another branch).

**Comment by AlexDaniel** [comment]
@kensanata what does returned `$revision` mean?
Here:

``` perl
sub GetTextRevision {
  my ($revision, $quiet) = @_;
  $revision =~ s/\D//g;   # Remove non-numeric chars
  return ($Page{text}, $revision, $Page{summary}) unless $revision and $revision ne $Page{revision};
  my %keep = GetKeptRevision($revision);
  if (not %keep) {
    $Message .= $q->p(Ts('Revision %s not available', $revision)
              . ' (' . T('showing current revision instead') . ')') unless $quiet;
    return ($Page{text}, '', '');
  }
  $Message .= $q->p(Ts('Showing revision %s', $revision)) unless $quiet;
  return ($keep{text}, $revision, $keep{summary});
}
```

We accept `$revision` as parameter (which revision should be returned), but when we return everything, `$revision` is returned back. Why?
So my naïve approach to return just the object is making a lot of tests unhappy. It looks like most of the time we are passing an empty string (`''`) but the real revision number is `1`. Currently it throws an empty string back, but I'm failing to understand the logic behind it.
Any ideas?

**Comment by kensanata** [comment]
I think it basically acts as a flag. If we return the _current version_ because we could not find the version requested, then we're also returning no revision.

**Comment by kensanata** [comment]
Hm…

```diff
alex@Megabombus:~/src/oddmuse (master *=)$ git diff
diff --git a/wiki.pl b/wiki.pl
index d0b13e3..fff972b 100755
--- a/wiki.pl
+++ b/wiki.pl
@@ -2722,15 +2722,15 @@ sub GetTextAtTime { # call with opened page, return $minor if all pages between
 sub GetTextRevision {
   my ($revision, $quiet) = @_;
   $revision =~ s/\D//g;   # Remove non-numeric chars
-  return ($Page{text}, $revision, $Page{summary}) unless $revision and $revision ne $Page{revision};
+  return ($Page{text}, $revision, $Page{summary}) if not $revision or $revision == $Page{revision};
   my %keep = GetKeptRevision($revision);
   if (not %keep) {
     $Message .= $q->p(Ts('Revision %s not available', $revision)
                      . ' (' . T('showing current revision instead') . ')') unless $quiet;
-    return ($Page{text}, '', '');
+    return ($Page{text}, '', ''); # empty means the current one
   }
   $Message .= $q->p(Ts('Showing revision %s', $revision)) unless $quiet;
-  return ($keep{text}, $revision, $keep{summary});
+  return ($keep{text}, $keep{revision}, $keep{summary});
 }
 
 sub GetPageContent {
```

**Comment by kensanata** [comment]
Assuming we were to return `\%keep` from `GetTextRevision` (perhaps rename it to `GetKeptPage`?) assigned the hash to `%Page` afterwards, we must still remember that this here is used to prevent edit conflicts:

```perl
  # shortcut if we only need the raw text: no caching, no diffs, no html.
  if ($raw) {
    print GetHttpHeader('text/plain', $Page{ts}, $IndexHash{$id} ? undef : '404 NOT FOUND');
    print $Page{ts} . " # Do not delete this line when editing!\n" if $raw == 2;
    print $text;
    return;
  }
```
And this here also expects `%Page` to refer to the `$OpenPageName`:

```perl
sub GetEditForm {
  my ($page_name, $upload, $oldText, $revision) = @_;
  my $html = GetFormStart(undef, undef, $upload ? 'edit upload' : 'edit text') # protected by questionasker
    .$q->p(GetHiddenValue("title", $page_name),
	   ($revision ? GetHiddenValue('revision', $revision) : ''),
           GetHiddenValue('oldtime', GetParam('oldtime', $Page{ts})), # prefer parameter over actual timestamp
	   ($upload ? GetUpload() : GetTextArea('text', $oldText)));
```
So, setting `%Page` will be tricky. Perhaps `GetTextRevision` should simply return one more element, `$keep{ts}`.

**Comment by AlexDaniel** [comment]
1c4e0827558d220fb624fe1af3245f4f386bee86 is a huge and painful commit that was done according to [Refactoring](https://oddmuse.org/wiki/Refactoring#returning_hashes).
This, however, does not really fix this problem (yet. But it is a step towards fixing it).
Now, here:

``` perl
sub DefaultFooter { # called via @MyFooters
  my ($id, $rev, $comment) = @_;
```

Uhm… it looks like we should pass our page object all the way through, right to the footer subs.
But until we start doing more changes, what do you think, @kensanata? Are we moving into the right direction? Or are we making our graves deeper?

**Comment by kensanata** [comment]
I think I like it. In general, I think we should try to get rid of globals like `%Page` and the like. This is old code from UseMod which was written when Perl 4 was still around. Once we eliminate globals, we'll be passing "our page object all the way through, right to the footer subs", right? I think we're moving in the right direction.

**Comment by AlexDaniel** [comment]
Yeah, exactly. Getting rid of `%Page` will eventually be possible.

**Comment by AlexDaniel** [comment]
One major problem is that page objects have no page name in them… that's unfortunate.

**Comment by kensanata** [comment]
We could add it as part of `OpenPage` and get rid of the `$OpenPageName` global. I think I'd like that.

**Comment by AlexDaniel** [comment]
@kensanata you are right, but for now let's just fix the problem. This is done in my recent commit. I've also added a test. Please merge this branch after you review the code (I did not really run this code in production yet, but all tests are fine and dandy). Also, I have added a note to [Upgrading Issues](https://oddmuse.org/wiki/Upgrading_Issues).

@phkehl you can try out [return-objects](https://github.com/kensanata/oddmuse/tree/return-objects) branch (but if you are going to update `wiki.pl`, make sure to update modules as well). If the problem is solved, feel free to close this issue. If you have any questions or suggestions, don't hesitate to write a comment. Thanks for reporting this bug!

**Comment by AlexDaniel** [comment]
Alex has just merged it into master (bf00a9ea04bf89b2db5a8ee08affbfb338890138).

**Comment by phkehl** [comment]
Dear Alex, Kensanata,

many thanks for fixing this, really appreciate it!

It works fine for me.
